### PR TITLE
[auth-nextjs] Async Request APIs, compat with Next.js 15

### PR DESCRIPTION
**Pull request by MartinCura**
This PR adapts the `@edgedb/auth-nextjs` package for the [kinda*] breaking change of Next.js 15, where [Request APIs like `cookies` and `params` are now async](https://nextjs.org/blog/next-15#async-request-apis-breaking-change).

To solve this, i simply added `await` to all corresponding places. Awaiting something sync is not strictly a noop, but for these cases it shouldn't change behavior afaict.

**Tested** in my own project with both Next.js before their breaking change and after it.

**Breaking change:** As `NextAppAuth.getSession()` has to await `cookies()`, it is now an `async` function, which means those upgrading have to add `await` when calling it.

```diff
const auth = createAuth(client, {
  baseUrl: `http://localhost:${process.env.PORT || 3000}`,
  authCookieName: "edgedb-session",
  pkceVerifierCookieName: "edgedb-pkce-verifier",
})

export const getCurrentUser = async () => {
- const session = auth.getSession()
+ const session = await auth.getSession()
  const userQuery = e.select(e.User, () => ({
    ...e.User["*"],
    filter_single: { id: e.global.current_user.id },
  }))
  return await userQuery.run(session.client)
}
```

This necessitates that `getSession` is now async, which is a breaking change, because it accesses the cookies.

*_kinda_ because as a compatibility layer it is still possible to not await it, but it's not what you want to do.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -24,5 +24,8 @@
     "format:fix": "prettier --write 'packages/*/(src|test)/**/*.(mts|ts)' 'integration-tests/*/*.(mts|ts)'",
     "ci:typecheck": "turbo run ci:typecheck",
     "ci:test": "turbo run ci:test"
+  },
+  "overrides": {
+    "@types/react": "npm:types-react@rc"
   }
```
Comment (MartinCura) [inline package.json:30]: this is recommended for the react types to be found correctly, at least while React 19 is not marked stable. for pnpm, this must be in the root of the workspace; i'm not actually sure if the same is true for yarn.

File: packages/auth-nextjs/src/app/index.ts [hunk 2]
```diff
@@ -27,18 +27,19 @@ export {
...
 
 export class NextAppAuth extends NextAuth {
-  getSession = cache(
-    () =>
-      new NextAuthSession(
-        this.client,
-        cookies().get(this.options.authCookieName)?.value.split(";")[0] ?? null,
-      ),
-  );
+  getSession = cache(async () => {
```
Comment (MartinCura) [inline packages/auth-nextjs/src/app/index.ts:30]: the necessary breaking change, making this function now async

File: packages/auth-nextjs/src/shared.ts [hunk 3]
```diff
@@ -93,6 +93,10 @@ export interface CreateAuthRouteHandlers {
   onSignout(req: NextRequest): Promise<never>;
 }
 
+type ReadonlyRequestCookies = Awaited<ReturnType<typeof nextCookies>>;
+const cookies: () => Promise<ReadonlyRequestCookies> | ReadonlyRequestCookies =
+  nextCookies;
```
Comment (MartinCura) [inline packages/auth-nextjs/src/shared.ts:98]: unnecessary, so can remove it, but just wanted to show that the return type depends on the next.js version used by the user

**Comment by 1st1** [comment]
Thank you, we'll take a look soon!

File: package.json
Comment (MartinCura) [inline package.json:1]: do you prefer i bump the minor version in this PR or do you prefer to do this in another one?

Comment (scotttrinh) [reply package.json:1]: We'll worry about the version numbers in the release PR, so no need to bump it here.

File: packages/auth-nextjs/src/shared.ts [hunk 4]
```diff
@@ -153,8 +159,9 @@ export abstract class NextAuth extends NextAuthHelpers {
     return {
       GET: async (
         req: NextRequest,
-        { params }: { params: { auth: string[] } },
+        ctx: { params: Promise<{ auth: string[] }> },
```
Comment (scotttrinh) [inline packages/auth-nextjs/src/shared.ts:162]: Will this work in <15 or will those users need to do a type-cast?

File: packages/auth-nextjs/src/shared.ts [hunk 5]
```diff
@@ -232,8 +239,9 @@ export abstract class NextAuth extends NextAuthHelpers {
                 req,
               );
             }
-            this.setAuthCookie(tokenData.auth_token);
-            cookies().delete(this.options.pkceVerifierCookieName);
+            await this.setAuthCookie(tokenData.auth_token);
+            const cookieStore = await cookies();
+            cookieStore.delete(this.options.pkceVerifierCookieName);
```
Comment (scotttrinh) [inline packages/auth-nextjs/src/shared.ts:244]: Not a blocker: Probably a good enough reason to promote this (and the auth cookie delete) to it's own function. Seemed superfluous when it was a one liner before.

**Review by scotttrinh — COMMENTED** [review]
Thanks so much for taking on this mechanical work! It's been on our "next up" list for a few weeks now.

File: packages/auth-nextjs/src/shared.ts (hunk 3 above)
Comment (scotttrinh) [reply packages/auth-nextjs/src/shared.ts:98]: Feel free to remove it in this PR, thanks for the demonstration!

File: packages/auth-nextjs/src/shared.ts (hunk 5 above)
Comment (MartinCura) [reply packages/auth-nextjs/src/shared.ts:244]: sgtm, i'll name it `deletePkceVerifierCookie()`

File: packages/auth-nextjs/src/shared.ts (hunk 4 above)
Comment (scotttrinh) [reply packages/auth-nextjs/src/shared.ts:162]: Yeah, I'm not worried too much about runtime, and even if it breaks backwards compatibility, I'm fine with shipping it, I just want to understand everything that might break here so we can update the docs.

File: packages/auth-nextjs/src/shared.ts (hunk 5 above)
Comment (MartinCura) [reply packages/auth-nextjs/src/shared.ts]: done

File: packages/auth-nextjs/src/shared.ts (hunk 4 above)
Comment (MartinCura) [reply packages/auth-nextjs/src/shared.ts:163]: Well the `GET` is exported in a `route.ts` to act as a Next.js Route Handler but it's not type checked against anything. I can run and build the project running Next.js before their breaking change with this version, no errors. And i don't know of any way that someone could be type-checking this and get a warning/error after this change.

At most we could add a small note to the docs referencing the change introduced by Next.js 15 and how this small library targets the new behaviour but supports both.

**Comment by MartinCura** [comment]
> Thanks so much for taking on this mechanical work! It's been on our "next up" list for a few weeks now.

No worries! Glad to help and love what you guys are doing. I'm working on a sizable project by myself, so these kinds of helper libraries that help me ship faster are very welcome.

**Comment by MartinCura** [comment]
👋🏼 Small update: React 19 stable finally dropped, and with it a stable release of `@types/react`, so the `overrides` i added in this PR should not be needed anymore.

**Pull request merged**